R/twas.R

```diff
@@ -92,19 +92,33 @@ twas_weights_cv <- function(X, Y, fold = NULL, sample_partitions = NULL, weight_
     # Create or use provided folds
     if (!is.null(fold)) {
         if (!is.null(seed)) set.seed(seed)
-        sample_indices <- sample(nrow(X))
-        folds <- cut(seq(1, nrow(X)), breaks = fold, labels = FALSE)
-        sample_partition <- data.frame(Sample = sample_names[sample_indices], Fold = folds, stringsAsFactors = FALSE)
+        
+        if (!is.null(sample_partitions)) {
+            if(fold!= max(unique(sample_partition$Fold))){
+                message(paste0("fold number provided does not match with sample partition, performing ", max(unique(sample_partition$Fold)),
+                       " fold cross validation based on provided sample partition. "))
+                }
+            
+            folds <- sample_partitions$Fold
+            sample_partition <- sample_partitions
+            
+        } else {
+            sample_indices <- sample(nrow(X))
+            folds <- cut(seq(1, nrow(X)), breaks = fold, labels = FALSE)
+            sample_partition <- data.frame(Sample = sample_names[sample_indices], Fold = folds, stringsAsFactors = FALSE)
+        }
     } else if (!is.null(sample_partitions)) {
         if (!all(sample_partitions$Sample %in% sample_names)) {
             stop("Some samples in 'sample_partitions' do not match the samples in 'X' and 'Y'.")
         }
         folds <- sample_partitions$Fold
         sample_partition <- sample_partitions
+        fold <- max(unique(sample_partition$Fold))
```

Comment:
It's safer to use `length(unique(sample_partition$Fold))`?

Reply:
ok, it is now fixed